modules/ui/sections/map_daterange.js

```diff
@@ -0,0 +1,167 @@
+];
+const LABEL_STYLES = [
+    { name: 'font-weight', value: 'bold' },
+    { name: 'display', value: 'inline-block' },
+    { name: 'width', value: '75px' },
+];
+
+export function uiSectionDateRange(context) {
+    // despite appearing as a separate panel, Map Features does the real filtering
+    // see applyDateRange() in this panel, where the dateRange value is set
+    // see modules/renderer/features.js checkDateFilter() which applies the filters
+    // see modules/renderer/features.js update() which updates URL params
+
+    const section = uiSection('date_ranges', context)
+        .label(() => t.append('date_ranges.title'))
+        .disclosureContent(renderDisclosureContent)
+        .expandedByDefault(false);
+
+    function renderDisclosureContent(selection) {
+        const container = selection.selectAll('.date_ranges-container').data([0]);
+
+        // start date label & input
+        const mindate_label = container.enter()
+            .append('label')
+            .html(t('date_ranges.start_date.description'))
+            .merge(container);
+        const mindate_input = container.enter()
+            .append('input')
+            .attr('type', 'text')
+            .attr('value', DEFAULT_MIN_DATE)
+            .attr('title', () => t.append('date_ranges.start_date.tooltip'))
+            .attr('placeholder', () => t.append('date_ranges.start_date.placeholder'))
+            .merge(container);
+
+        // line break
+        container.enter()
+            .append('br')
+            .merge(container);
+
+        // end date label & input
+        const maxdate_label = container.enter()
+            .append('label')
+            .html(t('date_ranges.end_date.description'))
+            .merge(container);
+        const maxdate_input = container.enter()  // we will refer to this widget by its name attribute to fetch our date range
+            .append('input')
+            .attr('type', 'text')
+            .attr('value', DEFAULT_MAX_DATE)
+            .attr('title', () => t.append('date_ranges.end_date.tooltip'))
+            .attr('placeholder', () => t.append('date_ranges.end_date.placeholder'))
+            .merge(container);
+
+        // apply styles
+        INPUT_STYLES.forEach(function (style) {
+            mindate_input.style(style.name, style.value);
+            maxdate_input.style(style.name, style.value);
+        });
+        LABEL_STYLES.forEach(function (style) {
+            mindate_label.style(style.name, style.value);
+            maxdate_label.style(style.name, style.value);
+        });
+
+        // event handler for change event
+        // intercept invalid & blank and correct them to our hardcoded in/max
+        // then cause a re-filter/redraw
+        function applyDateRange() {
+            const mindate = mindate_input.property('value');
+            const maxdate = maxdate_input.property('value');
+
+            context.features().dateRange = [mindate, maxdate];
+            context.flush();
```

Comment:
Flushing the context here causes OpenHistoricalMap/issues#365 and OpenHistoricalMap/issues#447. By comparison, the checkboxes under the Map Features section adjust the features’ hidden state but don’t toss any context:

https://github.com/OpenHistoricalMap/iD/blob/e9bdd79887c5c8272c3ac9d345aa10943adaf275/modules/ui/sections/map_features.js#L119-L121 https://github.com/OpenHistoricalMap/iD/blob/e9bdd79887c5c8272c3ac9d345aa10943adaf275/modules/renderer/features.js#L305-L310 https://github.com/OpenHistoricalMap/iD/blob/e9bdd79887c5c8272c3ac9d345aa10943adaf275/modules/renderer/features.js#L54-L69

If I’m not mistaken, it’s the `redrew` event that causes the canvas to redraw without the newly hidden features, but nothing ever changes in terms of the object model.

Reply:
Good call. I created a new `redraw()` wrapper in features.js, so `context.features().redraw();` now exists as an interface. The redrawing is MUCH FASTER now, and it sounds as if it will also correct some bugs. So that's great. Thank you.